File: runtime/java/src/main/java/dev/argon/verilization/runtime/I64.java
package dev.argon.verilization.runtime;

import java.io.IOException;

public final class I64 {
    private I64() {}

    public static long fromInteger(int i) {
        return i;
    }

    public static long fromInteger(long l) {
        return l;
    }

    public static final Codec<Long> codec = new Codec<Long>() {
        @Override
        public Long read(FormatReader reader) throws IOException {
            return reader.readLong();
        }

        @Override
        public void write(FormatWriter writer, Long value) throws IOException {
            writer.writeLong(value);
        }
    };
}

File: runtime/typescript/src/I64.ts
import { Codec } from "./Codec.js";
import { FormatReader, FormatWriter } from "./FormatIO.js";

export type I64 = bigint;

export const codec: Codec<I64> = {
    async read(reader: FormatReader): Promise<I64> {
        return BigInt.asIntN(64, await reader.readU64());
    },

    write(writer: FormatWriter, value: I64): Promise<void> {
        return writer.writeU64(BigInt.asUintN(64, value));
    },
};

export function fromInteger(n: bigint): I64 {
    return BigInt.asIntN(64, n);
}
